File: 318-maximum-product-of-word-lengths/maximum-product-of-word-lengths.java
class Solution {
    public int maxProduct(String[] words) {
        // bitmask -> max length of all strings with same bitmask
        Map<Integer, Integer> maskMap = new HashMap<>();
        for (String word: words) {
            int mask = 0;
            for (char c: word.toCharArray()) {
                mask = mask | (1 << (c - 'a'));
            }
            maskMap.put(mask, Math.max(maskMap.getOrDefault(mask, 0), word.length()));
        }

        int result = 0;
        for (int x: maskMap.keySet()) {
            for (int y: maskMap.keySet()) {
                if ((x & y) == 0) {
                    result = Math.max(result, maskMap.get(x) * maskMap.get(y));
                }
            }
        }
        return result;
    }
}

File: 318-maximum-product-of-word-lengths/maximum-product-of-word-lengths.py
class Solution:
    def maxProduct(self, words):
        # Dictionary to store bitmask -> max length of all strings with same bitmask
        mask_map = {}
        
        for word in words:
            mask = 0
            for c in word:
                mask |= 1 << (ord(c) - ord('a'))
            mask_map[mask] = max(mask_map.get(mask, 0), len(word))
        
        result = 0
        for x in mask_map:
            for y in mask_map:
                if x & y == 0:
                    result = max(result, mask_map[x] * mask_map[y])
        
        return result
